plugins/atomicfu/atomicfu-compiler/src/org.jetbrains.kotlinx.atomicfu.compiler/extensions/TransformerHelpers.kt

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlinx.atomicfu.compiler.extensions
+
+import org.jetbrains.kotlin.backend.common.extensions.IrPluginContext
+import org.jetbrains.kotlin.descriptors.ClassConstructorDescriptor
+import org.jetbrains.kotlin.descriptors.ClassDescriptor
+import org.jetbrains.kotlin.descriptors.FunctionDescriptor
+import org.jetbrains.kotlin.incremental.components.NoLookupLocation
+import org.jetbrains.kotlin.ir.UNDEFINED_OFFSET
+import org.jetbrains.kotlin.ir.backend.js.ir.JsIrBuilder
+import org.jetbrains.kotlin.ir.declarations.*
+import org.jetbrains.kotlin.ir.expressions.*
+import org.jetbrains.kotlin.ir.expressions.impl.IrCallImpl
+import org.jetbrains.kotlin.ir.expressions.impl.IrGetFieldImpl
+import org.jetbrains.kotlin.ir.expressions.impl.IrSetFieldImpl
+import org.jetbrains.kotlin.ir.symbols.*
+import org.jetbrains.kotlin.ir.types.*
+import org.jetbrains.kotlin.ir.types.impl.IrSimpleTypeImpl
+import org.jetbrains.kotlin.ir.types.impl.makeTypeProjection
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.name.Name
+import org.jetbrains.kotlin.resolve.scopes.MemberScope
+import org.jetbrains.kotlin.types.Variance
+import kotlin.random.Random
+
+interface TransformerHelpers {
+
+    val context: IrPluginContext
+
+    fun buildCall(
+        target: IrFunctionSymbol,
+        type: IrType? = null,
+        origin: IrStatementOrigin,
+        typeArguments: List<IrType>? = null,
+        valueArguments: List<IrExpression?>? = null
+    ): IrCall =
+        IrCallImpl(
+            UNDEFINED_OFFSET,
+            UNDEFINED_OFFSET,
+            type ?: target.owner.returnType,
+            target,
+            target.descriptor.typeParametersCount,
+            origin
+        ).apply {
+            typeArguments?.let {
+                assert(typeArguments.size == typeArgumentsCount)
+                it.withIndex().forEach { (i, t) -> putTypeArgument(i, t) }
+            }
+            valueArguments?.let {
+                assert(valueArguments.size == valueArgumentsCount)
+                it.withIndex().forEach { (i, arg) -> putValueArgument(i, arg) }
+            }
+        }
+
+    fun buildSetField(
+        symbol: IrFieldSymbol,
+        receiver: IrExpression?,
+        value: IrExpression,
+        superQualifierSymbol: IrClassSymbol? = null
+    ) =
+        IrSetFieldImpl(
+            UNDEFINED_OFFSET,
+            UNDEFINED_OFFSET,
+            symbol,
+            receiver,
+            value,
+            value.type,
+            IrStatementOrigin.INVOKE,
```

Comment:
`IrStatementOrigin.SET_PROPERTY`?

Reply:
We should add this origin to `IrStatementOrigin` Currently, there is only `IrStatementOrigin.GET_PROPERTY`.